### PR TITLE
CLAS: Fix several issues related to class includes

### LINKED ISSUE #4460: CLAS: Testclass include not removed
Reproduce:
1. Create a class with a test class (unit test)
2. Stage 
3. Remove test class on remote (for example, directly in GitHub: delete `zcl_test.clas.testclasses.abap` and set `<WITH_UNIT_TEST>` to initial in `zcl_test.clas.xml`)
4. Pull

As a result, the class will not have any unit tests but the test class include (CCAU) remains including the old code.

Recent examples:

image

Workaround: 

SE24 > Change Class > Utilities > Test Classes > Delete

image

### PR DESCRIPTION
**Pull request by mbtools**
There are several issues related to class includes (`locals_def, locals_imp, macros, testclasses`):

1. If unit tests are removed in a repo, the test class include was not removed during deserialization. The include might still contain obsolete code (#4460)
2. Similarly, if local definitions or implementations were deleted, the includes were not deleted during a pull (#4479)

The class includes mentioned above are now handled properly during a regular pull. A force pull should not be necessary anymore to resolve these situations. 

Closes #4460, #4479

Test repo to play around with. It's similar to the case mentioned in #1066 and contains branches with or with `locals_def` include:
https://github.com/marcfbe/TEST_RESET_LOCAL

Additional clean-up:
- removes temporary workaround #3833
- remove unused parameter
- replace some literals with constants
- moved add_to_activation_list

### CODE REVIEW

File: src/objects/zcl_abapgit_objects.clas.abap [hunk 1]
```diff
@@ -956,6 +954,15 @@ CLASS zcl_abapgit_objects IMPLEMENTATION.
       ENDLOOP.
     ENDIF.
 
+    "ignore objects that where deleted remotely
+    "but not class includes with separate file which are cleaned up during deserialize
+    LOOP AT rt_results REFERENCE INTO lr_object WHERE rstate = zif_abapgit_definitions=>c_state-deleted.
+      lv_tabix = sy-tabix.
+      IF zcl_abapgit_oo_base=>is_part_of_class( lr_object->filename ) = abap_false.
```
Comment (larshp) [inline src/objects/zcl_abapgit_objects.clas.abap:961]: this seems leaky, ideally, this method should be independent of the object type

is `lr_object->obj_type` and `lr_object->obj_name` filled with the correct information at this time?

File: src/ui/zcl_abapgit_services_git.clas.abap [hunk 2]
```diff
@@ -375,6 +375,11 @@ CLASS zcl_abapgit_services_git IMPLEMENTATION.
                       WHERE lstate = zif_abapgit_definitions=>c_state-added
                          OR rstate = zif_abapgit_definitions=>c_state-deleted.
 
+      " Class includes with separate files are cleaned up during deserialize
+      IF zcl_abapgit_oo_base=>is_part_of_class( <ls_status>-filename ) = abap_true.
```
Comment (larshp) [inline src/ui/zcl_abapgit_services_git.clas.abap:379]: see comment above

**Review by larshp — CHANGES_REQUESTED** [review]

File: src/objects/zcl_abapgit_objects.clas.abap (hunk 1 above)
Comment (mbtools) [reply src/objects/zcl_abapgit_objects.clas.abap:961]: yes, we could add `obj_type = 'CLAS'` to the 'if'. 

it's another case that should go on this list https://github.com/abapGit/abapGit/issues/4187 as `IS_PART_OF_OBJECT` or something like it.

**Comment by larshp** [comment]
I'd say, if the files for the remote object and the local object doesn't match, then the serializer for that object should be triggered during the pull?

**Comment by mbtools** [comment]
That's what the change does. Before, if for example locals_imp was deleted remotely, it didn't pull at all and with "force pull" you get the popup to confirm deletion of the complete class. 

After the change, the popup is not shown anymore (for the 4 class includes). The class is deserialized without deleting it first and deserialize cleans up the includes.

**Comment by larshp** [comment]
yea, it fixes the issue, but note that my comment talks about "objects" not specific CLAS'es or other object types

I think we will get in trouble if starting to add custom code for each object type on this level

**Comment by mbtools** [comment]
understood. let me think about it

File: src/objects/zcl_abapgit_objects.clas.abap (hunk 1 above)
Comment (mbtools) [reply src/objects/zcl_abapgit_objects.clas.abap]: I removed the is_part logic, so the behavior described in #1066 remains

**Comment by mbtools** [comment]
Removed critical parts and updated description (removed #1066)

File: src/ui/zcl_abapgit_services_git.clas.abap (hunk 2 above)
Comment (mbtools) [reply src/ui/zcl_abapgit_services_git.clas.abap]: removed

**Comment by mbtools** [comment]
tested with https://github.com/abapGit-tests/...
CLAS
CLAS_with_ENHO
CLAS_with_SOTR
CLAS_persistent
CLAS_textpool
CLAS_implementing_INTF
CLAS_full / main
CLAS_full / without_testclass
CLAS_APACK

all ok (after updating the test cases to latest AG to remove diffs)

if test class is removed in repo, a regular pull will remove it from the class (no more force pull required)

**Comment by larshp** [comment]
pending review by @larshp

**Comment by larshp** [comment]
unused parameter `iv_force` removed from method `generate_locals`

File: src/objects/zcl_abapgit_objects.clas.abap [hunk 3]
```diff
@@ -1346,7 +1346,9 @@ CLASS ZCL_ABAPGIT_OBJECTS IMPLEMENTATION.
     LOOP AT it_results ASSIGNING <ls_result> WHERE NOT obj_type IS INITIAL.
       IF <ls_result>-lstate IS NOT INITIAL
         AND NOT ( <ls_result>-lstate = zif_abapgit_definitions=>c_state-added
-        AND <ls_result>-rstate IS INITIAL ).
+        AND <ls_result>-rstate IS INITIAL )
+        OR ( <ls_result>-lstate IS INITIAL
+        AND <ls_result>-rstate = zif_abapgit_definitions=>c_state-deleted ).
```
Comment (larshp) [inline src/objects/zcl_abapgit_objects.clas.abap:1351]: nice and symmetric, so guess its okay

but remote and local state doesn't work properly(and never will), I see the overall logic to be changed to something simple like, if pulling, pull the objects where there are diffs(ignoring the local and remote calculations), but its a different story

LGTM

**Review by larshp — APPROVED** [review]

**Pull request merged**